statsmodels/tsa/statespace/kalman_filter.py

```diff
@@ -0,0 +1,830 @@
+
+                # For completely missing observations, the Kalman filter will
+                # produce forecasts, but forecast errors and the forecast
+                # error covariance matrix will be zeros - make them nan to
+                # improve clarity of results.
+                if self.nmissing[t] == self.k_endog:
+                    # We can recover forecasts
+                    self.forecasts[:, t] = np.dot(
+                        self.design[:, :, design_t], self.predicted_state[:, t]
+                    ) + self.obs_intercept[:, obs_intercept_t]
+                    self.forecasts_error[:, t] = np.nan
+                    self.forecasts_error_cov[:, :, t] = np.nan
+                # For partially missing observations, the Kalman filter
+                # will produce all elements (forecasts, forecast errors,
+                # forecast error covariance matrices) as usual, but their
+                # dimension will only be equal to the number of non-missing
+                # elements, and their location in memory will be in the first
+                # blocks (e.g. for the forecasts_error, the first
+                # k_endog - nmissing[t] columns will be filled in), regardless
+                # of which endogenous variables they refer to (i.e. the non-
+                # missing endogenous variables for that observation).
+                # Furthermore, the forecast error covariance matrix is only
+                # valid for those elements. What is done is to set all elements
+                # to nan for these observations so that they are flagged as
+                # missing. The variables missing_forecasts, etc. then provide
+                # the forecasts, etc. provided by the Kalman filter, from which
+                # the data can be retrieved if desired.
+                elif self.nmissing[t] > 0:
+                    self.forecasts[:, t] = np.nan
+                    self.forecasts_error[:, t] = np.nan
+                    self.forecasts_error_cov[:, :, t] = np.nan
+
+    @property
+    def kalman_gain(self):
+        if self._kalman_gain is None:
+            # k x n
+            self._kalman_gain = np.zeros(
+                (self.k_states, self.k_endog, self.nobs), dtype=self.dtype)
+            for t in range(self.nobs):
+                design_t = 0 if self.design.shape[2] == 1 else t
+                transition_t = 0 if self.transition.shape[2] == 1 else t
+                self._kalman_gain[:, :, t] = np.dot(
+                    np.dot(
+                        self.transition[:, :, transition_t],
+                        self.predicted_state_cov[:, :, t]
+                    ),
+                    np.dot(
+                        np.transpose(self.design[:, :, design_t]),
+                        np.linalg.inv(self.forecasts_error_cov[:, :, t])
+                    )
+                )
+        return self._kalman_gain
+
+    @property
+    def standardized_forecasts_error(self):
```

Comment:
Good point on the docstring. Since it already has a cache mechanism, I think I won't mess for now with `cache_readonly` in this class to keep options open for the future.